src/blockMonitor/blockchainMachine.ts

```diff
@@ -31,62 +106,92 @@ export class BlockchainMachine<TBlock extends IBlockStub> extends StartStopServi
      * @param component
      */
     public addComponent(component: Component<AnchorState, TBlock, ComponentAction>): void {
-        if (this.started) {
-            throw new ApplicationError("Components must be added before the BlockchainMachine is started.");
-        }
+        if (this.started) throw new ApplicationError("Components must be added before the BlockchainMachine is started.");
+        if (this.componentNames.has(component.name)) throw new ApplicationError(`A Component with the name "${component.name}" was already added.`);
 
-        this.componentsAndStates.push({
-            component,
-            states: new WeakMap()
-        });
+        this.components.push(component);
+        this.componentNames.add(component.name);
     }
 
-    private processNewBlock(block: TBlock) {
-        // Every time a new block is received we calculate the anchor state for that block and store it
+    private async processNewBlock(block: TBlock) {
+        try {
+            await this.lock.acquire();
 
-        for (const { component, states } of this.componentsAndStates) {
-            // If the parent is available and its anchor state is known, the state can be computed with the reducer.
-            // If the parent is available but its anchor state is not known, first compute its parent's initial state, then apply the reducer.
-            // Finally, if the parent is not available at all in the block cache, compute the initial state based on the current block.
+            // Every time a new block is received we calculate the anchor state for that block and store it
 
-            let newState: AnchorState;
-            if (this.blockProcessor.blockCache.hasBlock(block.parentHash)) {
-                const parentBlock = this.blockProcessor.blockCache.getBlock(block.parentHash);
-                const prevAnchorState = states.get(parentBlock) || component.reducer.getInitialState(parentBlock);
+            for (const component of this.components) {
+                // If the parent is available and its anchor state is known, the state can be computed with the reducer.
+                // If the parent is available but its anchor state is not known, first compute its parent's initial state, then apply the reducer.
+                // Finally, if the parent is not available at all in the block cache, compute the initial state based on the current block.
 
-                newState = component.reducer.reduce(prevAnchorState, block);
-            } else {
-                newState = component.reducer.getInitialState(block);
-            }
+                let newState: AnchorState;
+                let prevHeadAnchorState: AnchorState | null = null;
```

Comment:
Rather than having a nullable prevAnchorState, I'd be tempted to set prevAnchorState = newAnchorState where the parent block isnt available. Since the only case where this could happen is when PISA starts for the first time, and in that case the only consequence could only be to pass the newAnchorStae and prevAnchorState to detect changes, which should have negligible effect. If you cover/document this one edge case then you never need to worry about prevAnchorState being null.

Reply:
There are only two places where this matters, that is the new-block and the new-head event handlers, and there shouldn't be more in the future; hence leaving the nullability for now.